File: samples/book-app-project/books.py
import json
from dataclasses import dataclass, asdict
from typing import List, Optional

DATA_FILE = "data.json"
FORBIDDEN_INPUT_CHARS = {";", "|", "&", "`", "$", ">", "<", "\n", "\r"}


@dataclass
class Book:
    title: str
    author: str
    year: int
    read: bool = False


class BookCollection:
    def __init__(self):
        self.books: List[Book] = []
        self.load_books()

    def load_books(self):
        """Load books from the JSON file if it exists."""
        try:
            with open(DATA_FILE, "r") as f:
                data = json.load(f)
                self.books = [Book(**b) for b in data]
        except FileNotFoundError:
            self.books = []
        except json.JSONDecodeError:
            print("Warning: data.json is corrupted. Starting with empty collection.")
            self.books = []

    def save_books(self):
        """Save the current book collection to JSON."""
        with open(DATA_FILE, "w") as f:
            json.dump([asdict(b) for b in self.books], f, indent=2)

    @staticmethod
    def _validate_text_input(value: str, field_name: str) -> str:
        if not isinstance(value, str):
            raise ValueError(f"{field_name} must be a string.")

        cleaned_value = value.strip()
        if len(cleaned_value) > 200:
            raise ValueError(f"{field_name} is too long.")
        if any(char in FORBIDDEN_INPUT_CHARS for char in cleaned_value):
            raise ValueError(f"{field_name} contains forbidden characters.")

        return cleaned_value

    @staticmethod
    def _validate_year(year: int) -> int:
        if not isinstance(year, int):
            raise ValueError("Year must be an integer.")
        if year < 0 or year > 9999:
            raise ValueError("Year must be between 0 and 9999.")
        return year

    def add_book(self, title: str, author: str, year: int) -> Book:
        title = self._validate_text_input(title, "Title")
        author = self._validate_text_input(author, "Author")
        year = self._validate_year(year)
        book = Book(title=title, author=author, year=year)
        self.books.append(book)
        self.save_books()
        return book

    def list_books(self) -> List[Book]:
        return self.books

    def find_book_by_title(self, title: str) -> Optional[Book]:
        title = self._validate_text_input(title, "Title")
        for book in self.books:
            if book.title.lower() == title.lower():
                return book
        return None

    def mark_as_read(self, title: str) -> bool:
        book = self.find_book_by_title(title)
        if book:
            book.read = True
            self.save_books()
            return True
        return False

    def remove_book(self, title: str) -> bool:
        """Remove a book by title."""
        title = self._validate_text_input(title, "Title")
        book = self.find_book_by_title(title)
        if book:
            self.books.remove(book)
            self.save_books()
            return True
        return False

    def find_by_author(self, author: str) -> List[Book]:
        """Find all books by a given author."""
        author = self._validate_text_input(author, "Author")
        return [b for b in self.books if b.author.lower() == author.lower()]

File: samples/book-app-project-js/books.js
const fs = require("fs");
const path = require("path");

const DATA_FILE = path.join(__dirname, "data.json");
const FORBIDDEN_INPUT_CHARS = /[;|&`$><\n\r]/;

class Book {
  constructor(title, author, year, read = false) {
    this.title = title;
    this.author = author;
    this.year = year;
    this.read = read;
  }
}

class BookCollection {
  constructor(dataFile) {
    this.dataFile = dataFile || DATA_FILE;
    this.books = [];
    this.loadBooks();
  }

  loadBooks() {
    try {
      const raw = fs.readFileSync(this.dataFile, "utf-8");
      const data = JSON.parse(raw);
      this.books = data.map((b) => new Book(b.title, b.author, b.year, b.read));
    } catch (err) {
      if (err.code === "ENOENT") {
        this.books = [];
      } else if (err instanceof SyntaxError) {
        console.log("Warning: data.json is corrupted. Starting with empty collection.");
        this.books = [];
      } else {
        throw err;
      }
    }
  }

  saveBooks() {
    const data = this.books.map((b) => ({
      title: b.title,
      author: b.author,
      year: b.year,
      read: b.read,
    }));
    fs.writeFileSync(this.dataFile, JSON.stringify(data, null, 2));
  }

  static validateTextInput(value, fieldName) {
    if (typeof value !== "string") {
      throw new Error(`${fieldName} must be a string.`);
    }

    const cleanedValue = value.trim();
    if (cleanedValue.length > 200) {
      throw new Error(`${fieldName} is too long.`);
    }
    if (FORBIDDEN_INPUT_CHARS.test(cleanedValue)) {
      throw new Error(`${fieldName} contains forbidden characters.`);
    }

    return cleanedValue;
  }

  static validateYear(year) {
    if (!Number.isInteger(year)) {
      throw new Error("Year must be an integer.");
    }
    if (year < 0 || year > 9999) {
      throw new Error("Year must be between 0 and 9999.");
    }
    return year;
  }

  addBook(title, author, year) {
    const safeTitle = BookCollection.validateTextInput(title, "Title");
    const safeAuthor = BookCollection.validateTextInput(author, "Author");
    const safeYear = BookCollection.validateYear(year);
    const book = new Book(safeTitle, safeAuthor, safeYear);
    this.books.push(book);
    this.saveBooks();
    return book;
  }

  listBooks() {
    return this.books;
  }

  findBookByTitle(title) {
    const safeTitle = BookCollection.validateTextInput(title, "Title");
    return this.books.find((b) => b.title.toLowerCase() === safeTitle.toLowerCase()) || null;
  }

  markAsRead(title) {
    const book = this.findBookByTitle(title);
    if (book) {
      book.read = true;
      this.saveBooks();
      return true;
    }
    return false;
  }

  removeBook(title) {
    const book = this.findBookByTitle(title);
    if (book) {
      this.books = this.books.filter((b) => b !== book);
      this.saveBooks();
      return true;
    }
    return false;
  }

  findByAuthor(author) {
    const safeAuthor = BookCollection.validateTextInput(author, "Author");
    return this.books.filter((b) => b.author.toLowerCase() === safeAuthor.toLowerCase());
  }
}

module.exports = { Book, BookCollection, DATA_FILE };
